File: v1/stringsArrays/PrimeNumbers.java
/**
 * Prime numbers (sieve of eratosthenes):
 * Print all prime numbers upto a given number
 * OR
 * Leetcode version: Count the number of prime numbers less than a non-negative number, n.
 */

/*
 * time: O(n log log n)
 * space: O(n)
 */

import java.util.*;

public class PrimeNumbers {
    public int countPrimes(int n) {
        if (n < 2) {
            return 0;
        }

        int[] arr = new int[n];
        int sqrt = (int) Math.sqrt(n);
        int count = n - 2;
        for (int i = 2; i <= sqrt; ++i) {
            if (arr[i] == 1) {
                continue;
            }

            for (int multiple = 2 * i; multiple < n; multiple += i) {
                if (arr[multiple] == 0) {
                    arr[multiple] = 1;
                    --count;
                }
            }
        }

        return count;
    }

    public static void main(String[] args) {
        PrimeNumbers pn = new PrimeNumbers();
        int result = pn.countPrimes(Integer.parseInt(args[0]));
        System.out.println("result: " + result);      
    }
}

File: primeNumbers.cpp
/* Sieve of Eratosthenes: Print prime numbers not greater than the given value n */

#include <iostream>
#include <vector>
#include <math.h>

using namespace std;

void printPrimeNumbers(int n)
{
	vector<bool> a(n + 1, true);
	
	int sqrtN = (int) ceil(sqrt(n));
	for (int i = 2; i <= sqrtN; ++i) {
		if (a[i] == false) {
			continue;
		}
		
		int sqI = pow(i, 2);
		int count = 0;
		for (int j = sqI; j <= n; j += i) {
			a[j] = false;
		}
	}
	
	cout << "Prime numbers: " << endl;
	for (int i = 2; i <= n; ++i) {
		if (a[i] == true) {
			cout << i << " ";
		}
	}
	cout << endl;
}

int main(int argc, char **argv)
{
	if (argc < 2) {
		cout << "Usage: a.out <max num>" << endl;
		return 1;
	} 
	
	int n = atoi(argv[1]);
	if (n <= 1) {
		cout << "The max number should be > 1" << endl;
		return 1;
	}
	
	printPrimeNumbers(n);
	
	return 0;
}